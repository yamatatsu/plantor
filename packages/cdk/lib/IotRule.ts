import {
  App,
  Stack,
  StackProps,
  aws_iot,
  RemovalPolicy,
  aws_logs,
  aws_iam,
  aws_s3,
} from "aws-cdk-lib";

type Props = StackProps & {};

export class IotRule extends Stack {
  constructor(parent: App, id: string, props: Props) {
    super(parent, id, props);

    const bucket = new aws_s3.Bucket(this, "Bucket", {
      removalPolicy: RemovalPolicy.DESTROY,
      blockPublicAccess: aws_s3.BlockPublicAccess.BLOCK_ALL,
    });

    const logGroup = new aws_logs.LogGroup(this, "MyLogGroup", {
      removalPolicy: RemovalPolicy.DESTROY,
    });

    const role = new aws_iam.Role(this, "Role", {
      assumedBy: new aws_iam.ServicePrincipal("iot.amazonaws.com"),
    });
    bucket.grantPut(role, "*");
    role.addToPrincipalPolicy(
      new aws_iam.PolicyStatement({
        actions: ["cloudwatch:PutMetricData"],
        resources: ["*"],
      }),
    );
    logGroup.grant(
      role,
      "logs:CreateLogStream",
      "logs:DescribeLogStreams",
      "logs:PutLogEvents",
    );

    new aws_iot.CfnTopicRule(this, "moisture_rule", {
      ruleName: "moisture_rule",
      topicRulePayload: {
        awsIotSqlVersion: "2016-03-23",
        sql: q(`SELECT
                  topic() as topic,
                  topic(2) + topic(3) as sensorName,
                  timestamp() as timestamp,
                  *
                FROM
                  'plantor/moisture/+'`),
        actions: [
          {
            s3: {
              bucketName: bucket.bucketName,
              key: "${topic(2) + topic(3)}/${timestamp()}.json",
              roleArn: role.roleArn,
            },
          },
          {
            cloudwatchMetric: {
              metricName: "${topic(2) + topic(3)}",
              metricNamespace: "CUSTOM-IoT/Moisture",
              metricTimestamp: "${timestamp() / 1000}",
              metricUnit: "None",
              metricValue: "${moisture}",
              roleArn: role.roleArn,
            },
          },
        ],
        errorAction: {
          cloudwatchLogs: {
            logGroupName: logGroup.logGroupName,
            roleArn: role.roleArn,
          },
        },
      },
    });
  }
}

const q = (query: string) => query.replace(/\n/g, " ").replace(/ +/g, " ");
