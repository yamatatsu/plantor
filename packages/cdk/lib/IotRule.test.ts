import { SynthUtils } from "@aws-cdk/assert";
import { App } from "aws-cdk-lib";
import { IotRule } from "./IotRule";

test("snapshot test", () => {
  const app = new App();

  const target = new IotRule(app, "Target", {});

  expect(SynthUtils.toCloudFormation(target)).toMatchSnapshot();
});
