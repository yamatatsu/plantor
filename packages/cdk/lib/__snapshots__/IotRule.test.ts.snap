// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test 1`] = `
Object {
  "Resources": Object {
    "Bucket83908E77": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "PublicAccessBlockConfiguration": Object {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
    },
    "MyLogGroup5C0DAD85": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "RetentionInDays": 731,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Delete",
    },
    "Role1ABCC5F0": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "iot.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "RoleDefaultPolicy5FFB7DAB": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:PutObject",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    Object {
                      "Fn::GetAtt": Array [
                        "Bucket83908E77",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
            Object {
              "Action": "cloudwatch:PutMetricData",
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "logs:CreateLogStream",
                "logs:DescribeLogStreams",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "MyLogGroup5C0DAD85",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "RoleDefaultPolicy5FFB7DAB",
        "Roles": Array [
          Object {
            "Ref": "Role1ABCC5F0",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "moisturerule": Object {
      "Properties": Object {
        "RuleName": "moisture_rule",
        "TopicRulePayload": Object {
          "Actions": Array [
            Object {
              "S3": Object {
                "BucketName": Object {
                  "Ref": "Bucket83908E77",
                },
                "Key": "\${topic()}/\${timestamp()}",
                "RoleArn": Object {
                  "Fn::GetAtt": Array [
                    "Role1ABCC5F0",
                    "Arn",
                  ],
                },
              },
            },
            Object {
              "CloudwatchMetric": Object {
                "MetricName": "001",
                "MetricNamespace": "CUSTOM-IoT/Moisture",
                "MetricTimestamp": "\${timestamp() / 1000}",
                "MetricUnit": "None",
                "MetricValue": "\${moisture}",
                "RoleArn": Object {
                  "Fn::GetAtt": Array [
                    "Role1ABCC5F0",
                    "Arn",
                  ],
                },
              },
            },
          ],
          "AwsIotSqlVersion": "2016-03-23",
          "ErrorAction": Object {
            "CloudwatchLogs": Object {
              "LogGroupName": Object {
                "Ref": "MyLogGroup5C0DAD85",
              },
              "RoleArn": Object {
                "Fn::GetAtt": Array [
                  "Role1ABCC5F0",
                  "Arn",
                ],
              },
            },
          },
          "Sql": "SELECT moisture, timestamp() as timestamp, * FROM 'wio_terminal/moisture'",
        },
      },
      "Type": "AWS::IoT::TopicRule",
    },
  },
}
`;
